src/main/java/org/jenkinsci/plugins/workflow/graph/StandardGraphLookupView.java

```diff
@@ -31,13 +31,17 @@
     HashMap<String, String> nearestEnclosingBlock = new HashMap<>();
 
     public void clearCache() {
-        blockStartToEnd.clear();
-        nearestEnclosingBlock.clear();
+        synchronized (blockStartToEnd) {
+            blockStartToEnd.clear();
+        }
+        synchronized (nearestEnclosingBlock) {
+            nearestEnclosingBlock.clear();
+        }
     }
 
     /** Update with a new node added to the flowgraph */
     @Override
-    public void onNewHead(@Nonnull FlowNode newHead) {
+    public synchronized void onNewHead(@Nonnull FlowNode newHead) {
```

Comment:
I preferred to `synchronized` the whole method as it uses both `blockStartToEnd` and `nearestEnclosingBlock` and the paths in the method are pretty simple. I thought it would be more readable this way as synchronising each maps at the proper use wouldn't have a huge added value.

Reply:
but as us this is still no longer safe?
thread 1 could call `bruteForceScanForEnd(..)` or `getEndNode()` and thread2 `onNewHead(...)`  resulting in concurrent search and modification which is IIUC the exact thing you are trying to defend against?

Reply:
Yes, you're right. I'm going to change that. Thanks

Reply:
**Resolve conversation**?